balancer-js/src/modules/liquidity-managment/migrations.integrations.spec.ts

```diff
@@ -28,296 +29,361 @@ describe('Migrations', function () {
     let signer: JsonRpcSigner;
     let address: string;
 
-    const migrations = new Migrations(
+    const migrations = new Migrations({
       relayerAddress,
-      poolRepository,
+      poolsRepository,
       gaugesRepository,
-      provider
-    );
+      provider,
+    });
+
+    context('user did not approve the relayer', () => {
+      let authorisation = '';
+
+      before(async () => {
+        await reset('https://rpc.ankr.com/eth', provider, 16950000);
+        signer = provider.getSigner();
+        address = await signer.getAddress();
+
+        await setTokenBalance(
+          signer,
+          '0xa6468eca7633246dcb24e5599681767d27d1f978',
+          1,
+          '1000000000000000000',
+          true
+        );
+        authorisation = await Relayer.signRelayerApproval(
+          relayerAddress,
+          address,
+          signer,
+          vault
+        );
+      });
+
+      it('should build gauge2gauge with signed authorisation', async () => {
+        const from = '0xa6468eca7633246dcb24e5599681767d27d1f978';
+        const to = '0x57ab3b673878c3feab7f8ff434c40ab004408c4c';
+        const balance = (
+          await ERC20__factory.connect(from, provider).balanceOf(address)
+        ).toString();
+
+        const txParams = await migrations.gauge2gauge({
+          user: address,
+          from,
+          to,
+          balance,
+          authorisation,
+        });
+
+        await (await signer.sendTransaction(txParams)).wait();
 
-    beforeEach(async () => {
-      await reset('https://rpc.ankr.com/eth', provider, 16950000);
-      signer = await impersonateAccount(address, provider);
+        const balanceAfter = (
+          await ERC20__factory.connect(to, provider).balanceOf(address)
+        ).toString();
 
-      // approve relayer
-      await vault
-        .connect(signer)
-        .setRelayerApproval(address, relayerAddress, true);
+        expect(balanceAfter).to.be.eql(balance);
+      });
     });
 
-    context('Metastable to Metastable', () => {
-      const from = metaStable;
-      const to = from;
+    context('user approved the relayer', () => {
+      beforeEach(async () => {
+        await reset('https://rpc.ankr.com/eth', provider, 16950000);
+        signer = await impersonateAccount(address, provider);
+
+        // approve relayer
+        await vault
+          .connect(signer)
+          .setRelayerApproval(address, relayerAddress, true);
+      });
+
+      context('Metastable to Metastable', () => {
+        const from = metaStable;
+        const to = from;
+
+        describe('bptHodler', () => {
```

Comment:
I feel like we could have a lot of benefits by extracting these tests to a reusable testFlow, since they all have the same structure:
- get balance before
- build peek transaction
- peek result using staticCall
- build/send transaction
- get balance after
- check results

What do you think?

Reply:
I was following this PR cause I was interested in `gauge2gauge` implementation. In my experience abusing `beforeEach/before` hooks makes the tests more difficult to read for a new developer. I prefer to be more explicit and extracting reusable functions that can be explicitly used for any of the individual tests. 

In this example, we can extract repeated things to functions such as: `getBalance`, `sendTransaction` or `peekBptOut`, leading tests that are easier to read but also to maintain (if you have edge cases like @gmbronco pointed out). 

```typescript
// yarn test:only ./src/modules/liquidity-managment/migrations.integrations.spec.ts
import { impersonateAccount, reset, setTokenBalance } from '@/test/lib/utils';
import { expect } from 'chai';
import { ERC20__factory, Vault__factory } from '@/contracts';
import { BALANCER_NETWORK_CONFIG } from '@/lib/constants/config';
import { JsonRpcProvider, JsonRpcSigner } from '@ethersproject/providers';
import {
  vitaDao1,
  vitaDao2,
  metaStable,
  composableStable,
  poolsRepository,
  gaugesRepository,
  polygonComposableStable,
  polygonPoolRepository,
} from './migrations/builder.spec-helpers';
import { Migrations } from './migrations';
import { Relayer } from '@/.';

const provider = new JsonRpcProvider('http://127.0.0.1:8545');

type TransactionParams = {
  to: string;
  data: string;
};

async function getBalance(from: string, address: string) {
  const balance = (
    await ERC20__factory.connect(from, provider).balanceOf(address)
  ).toString();
  return balance;
}

async function sendTransaction(
  signer: JsonRpcSigner,
  txParams: TransactionParams
) {
  await (await signer.sendTransaction(txParams)).wait();
}

async function peekBptOut(
  signer: JsonRpcSigner,
  peek: TransactionParams
): Promise<string> {
  const peekResult = await signer.call({ ...peek, gasLimit: 8e6 });
  const expectedBptOut = Migrations.getMinBptOut(peekResult);
  return expectedBptOut;
}

describe('Migrations', function () {
  this.timeout(30000);
  context('mainnet', () => {
    const {
      addresses: { contracts },
    } = BALANCER_NETWORK_CONFIG[1];
    const relayerAddress = contracts.relayer;
    const provider = new JsonRpcProvider('http://127.0.0.1:8545');
    const vault = Vault__factory.connect(contracts.vault, provider);
    let signer: JsonRpcSigner;
    let address: string;

    const migrations = new Migrations({
      relayerAddress,
      poolsRepository,
      gaugesRepository,
      provider,
    });

    context('user did not approve the relayer', () => {
      let authorisation = '';

      before(async () => {
        await reset('https://rpc.ankr.com/eth', provider, 16950000);
        signer = provider.getSigner();
        address = await signer.getAddress();

        await setTokenBalance(
          signer,
          '0xa6468eca7633246dcb24e5599681767d27d1f978',
          1,
          '1000000000000000000',
          true
        );
        authorisation = await Relayer.signRelayerApproval(
          relayerAddress,
          address,
          signer,
          vault
        );
      });

      it('should build gauge2gauge with signed authorisation', async () => {
        const from = '0xa6468eca7633246dcb24e5599681767d27d1f978';
        const to = '0x57ab3b673878c3feab7f8ff434c40ab004408c4c';
        const balance = await getBalance(from, to);

        const txParams = await migrations.gauge2gauge({
          user: address,
          from,
          to,
          balance,
          authorisation,
        });

        await sendTransaction(signer, txParams);

        const balanceAfter = await getBalance(to, address);

        expect(balanceAfter).to.be.eql(balance);
      });
    });

    context('user approved the relayer', () => {
      beforeEach(async () => {
        await reset('https://rpc.ankr.com/eth', provider, 16950000);
        signer = await impersonateAccount(address, provider);

        // approve relayer
        await vault
          .connect(signer)
          .setRelayerApproval(address, relayerAddress, true);
      });

      context('Metastable to Metastable', () => {
        const from = metaStable;
        const to = from;

        describe('bptHodler', () => {
          before(() => {
            address = '0x21ac89788d52070D23B8EaCEcBD3Dc544178DC60';
          });

          it('joins a new pool with an limit', async () => {
            const balance = await getBalance(from.address, address);

            const peek = await migrations.pool2pool({
              user: address,
              from: from.id,
              to: to.id,
              balance,
            });
            const expectedBptOut = await peekBptOut(signer, peek);

            const txParams = await migrations.pool2pool({
              user: address,
              from: from.id,
              to: to.id,
              balance,
              minBptOut: expectedBptOut,
            });

            await sendTransaction(signer, txParams);

            const balanceAfter = await getBalance(to.address, address);

            expect(balanceAfter).to.be.eq(expectedBptOut);
          });
        });

        describe('staked bpt', () => {
          before(() => {
            address = '0xe8343fd029561289CF7359175EE84DA121817C71';
          });

          it('should build a migration using exit / join and stake tokens in the gauge', async () => {
            const gauge = (await gaugesRepository.findBy(
              'poolId',
              from.id
            )) as {
              id: string;
            };
            const balance = await getBalance(gauge.id, address);

            const peek = await migrations.pool2poolWithGauges({
              user: address,
              from: from.id,
              to: to.id,
              balance,
            });
            const expectedBptOut = await peekBptOut(signer, peek);

            const txParams = await migrations.pool2poolWithGauges({
              user: address,
              from: from.id,
              to: to.id,
              balance,
              minBptOut: expectedBptOut,
            });

            await sendTransaction(signer, txParams);

            const balanceAfter = await getBalance(gauge.id, address);

            expect(balanceAfter).to.be.eq(expectedBptOut);
          });
        });
      });

      context('ComposableStable to ComposableStable', () => {
        before(() => {
          address = '0x74C3646ADad7e196102D1fE35267aDFD401A568b';
        });

        it('should build a migration using exit / join', async () => {
          const pool = composableStable;
          const balance = await getBalance(pool.address, address);

          const peek = await migrations.pool2pool({
            user: address,
            from: pool.id,
            to: pool.id,
            balance,
          });
          const expectedBptOut = await peekBptOut(signer, peek);

          // NOTICE: When swapping from Linear Pools, the swap will query for the current wrapped token rate.
          // It is possible that the rate changes between the static call checking for the BPT out
          // and the actual swap, causing it to fail with BAL#208.
          // To avoid this, we can add a small buffer to the min BPT out amount. eg. 0.0000001% of the BPT amount.
          const buffer = BigInt(expectedBptOut) / BigInt(1e9);
          const minBptOut = String(BigInt(expectedBptOut) - buffer);

          const txParams = await migrations.pool2pool({
            user: address,
            from: pool.id,
            to: pool.id,
            balance,
            minBptOut,
          });

          await sendTransaction(signer, txParams);

          const balanceAfter = await getBalance(pool.address, address);

          // NOTICE: We don't know the exact amount of BPT that will be minted,
          // because swaps from the linear pool are not deterministic due to external rates
          expect(BigInt(balanceAfter)).to.satisfy(
            (v: bigint) => v > v - v / buffer && v < v + v / buffer
          );
        });
      });

      context('Weighted to Weighted between different pools', () => {
        before(() => {
          address = '0x673CA7d2faEB3c02c4cDB9383344ae5c9738945e';
        });

        it('should build a migration using exit / join', async () => {
          const from = vitaDao1;
          const to = vitaDao2;
          const balance = await getBalance(from.address, to.address);

          const peek = await migrations.pool2pool({
            user: address,
            from: from.id,
            to: to.id,
            balance,
          });
          const expectedBptOut = await peekBptOut(signer, peek);

          const txParams = await migrations.pool2pool({
            user: address,
            from: from.id,
            to: to.id,
            balance,
            minBptOut: expectedBptOut,
          });

          await sendTransaction(signer, txParams);

          const balanceAfter = await getBalance(to.address, address);

          expect(balanceAfter).to.be.eq(expectedBptOut);
        });
      });

      context('gauge to gauge', () => {
        before(() => {
          address = '0xaF297deC752c909092A117A932A8cA4AaaFF9795';
        });

        it('should build a migration using exit / join and stake tokens in the gauge', async () => {
          const from = '0xa6468eca7633246dcb24e5599681767d27d1f978';
          const to = '0x57ab3b673878c3feab7f8ff434c40ab004408c4c';
          const balance = await getBalance(from, to);

          const txParams = await migrations.gauge2gauge({
            user: address,
            from,
            to,
            balance,
          });

          await sendTransaction(signer, txParams);

          const balanceAfter = await getBalance(from, to);

          expect(balanceAfter).to.be.eql(balance);
        });
      });
    });

    context('polygon', () => {
      const {
        addresses: { contracts },
      } = BALANCER_NETWORK_CONFIG[137];
      const relayerAddress = contracts.relayer;
      const provider = new JsonRpcProvider('http://127.0.0.1:8137');
      const vault = Vault__factory.connect(contracts.vault, provider);
      let signer: JsonRpcSigner;
      let address: string;

      const migrations = new Migrations({
        relayerAddress,
        poolsRepository: polygonPoolRepository,
        gaugesRepository,
        provider,
      });

      beforeEach(async () => {
        await reset('https://rpc.ankr.com/polygon', provider, 42462957);
        signer = await impersonateAccount(address, provider);

        // approve relayer
        await vault
          .connect(signer)
          .setRelayerApproval(address, relayerAddress, true);
      });

      context('ComposableStable to ComposableStable', () => {
        before(() => {
          address = '0xe80a6a7b4fdadf0aa59f3f669a8d394d1d4da86b';
        });

        it.skip('should build a migration using exit / join', async () => {
          const pool = polygonComposableStable;
          const balance = await getBalance(pool.address, address);

          const peek = await migrations.pool2pool({
            user: address,
            from: pool.id,
            to: pool.id,
            balance,
          });
          const expectedBptOut = await peekBptOut(signer, peek);

          // NOTICE: When swapping from Linear Pools, the swap will query for the current wrapped token rate.
          // It is possible that the rate changes between the static call checking for the BPT out
          // and the actual swap, causing it to fail with BAL#208.
          // To avoid this, we can add a small buffer to the min BPT out amount. eg. 0.0000001% of the BPT amount.
          const buffer = BigInt(expectedBptOut) / BigInt(1e14); // 0.0000001%
          const minBptOut = String(BigInt(expectedBptOut) - buffer);

          const txParams = await migrations.pool2pool({
            user: address,
            from: pool.id,
            to: pool.id,
            balance,
            minBptOut,
          });

          await sendTransaction(signer, txParams);

          const balanceAfter = await getBalance(pool.address, address);

          // NOTICE: We don't know the exact amount of BPT that will be minted,
          // because swaps from the linear pool are not deterministic due to external rates
          expect(BigInt(balanceAfter)).to.satisfy(
            (v: bigint) => v > v - buffer && v < v + buffer
          );
        });
      });
    });
  });
});
```

Reply:
There's still some repetition but I don't think it is so bad. In case it got worst, instead of going to something super generic like `testFlow` I would go for extracting semantic functions that can be reusable and easily maintainable.

In this example, we can extract  common functions such as `peekPool2Pool, runPool2Pool, runGauge2Gauge` leading to: 

```typescript
// yarn test:only ./src/modules/liquidity-managment/migrations.integrations.spec.ts
import { impersonateAccount, reset, setTokenBalance } from '@/test/lib/utils';
import { expect } from 'chai';
import { ERC20__factory, Vault__factory } from '@/contracts';
import { BALANCER_NETWORK_CONFIG } from '@/lib/constants/config';
import { JsonRpcProvider, JsonRpcSigner } from '@ethersproject/providers';
import {
  vitaDao1,
  vitaDao2,
  metaStable,
  composableStable,
  poolsRepository,
  gaugesRepository,
  polygonComposableStable,
  polygonPoolRepository,
} from './migrations/builder.spec-helpers';
import { Migrations } from './migrations';
import { Pool, Relayer } from '@/.';

const provider = new JsonRpcProvider('http://127.0.0.1:8545');

type TransactionParams = {
  to: string;
  data: string;
};

async function getBalance(from: string, address: string) {
  const balance = (
    await ERC20__factory.connect(from, provider).balanceOf(address)
  ).toString();
  return balance;
}

async function sendTransaction(
  signer: JsonRpcSigner,
  txParams: TransactionParams
) {
  await (await signer.sendTransaction(txParams)).wait();
}

async function peekBptOut(
  signer: JsonRpcSigner,
  peek: TransactionParams
): Promise<string> {
  const peekResult = await signer.call({ ...peek, gasLimit: 8e6 });
  const expectedBptOut = Migrations.getMinBptOut(peekResult);
  return expectedBptOut;
}

function setupMigrations(migrations: Migrations, signer: JsonRpcSigner) {
  return { peekPool2Pool, runPool2Pool, runGauge2Gauge };

  async function peekPool2Pool(user: string, from: Pool, to: Pool) {
    const balance = await getBalance(from.address, user);

    const peek = await migrations.pool2pool({
      user: user,
      from: from.id,
      to: to.id,
      balance,
    });

    const expectedBptOut = await peekBptOut(signer, peek);
    return { expectedBptOut, balance, runGauge2Gauge };
  }

  async function runPool2Pool(
    user: string,
    from: Pool,
    to: Pool,
    balance: string,
    expectedBptOut: string
  ) {
    const txParams = await migrations.pool2pool({
      user,
      from: from.id,
      to: to.id,
      balance,
      minBptOut: expectedBptOut,
    });

    await sendTransaction(signer, txParams);

    return await getBalance(to.address, user);
  }

  async function runGauge2Gauge(
    user: string,
    from: string,
    to: string,
    balance: string,
    authorisation?: string
  ) {
    const txParams = await migrations.gauge2gauge({
      user,
      from,
      to,
      balance,
      authorisation,
    });

    await sendTransaction(signer, txParams);

    return await getBalance(to, user);
  }
}

describe('Migrations', function () {
  this.timeout(30000);
  context('mainnet', () => {
    const {
      addresses: { contracts },
    } = BALANCER_NETWORK_CONFIG[1];
    const relayerAddress = contracts.relayer;
    const provider = new JsonRpcProvider('http://127.0.0.1:8545');
    const vault = Vault__factory.connect(contracts.vault, provider);
    let signer: JsonRpcSigner;
    let address: string;

    const migrations = new Migrations({
      relayerAddress,
      poolsRepository,
      gaugesRepository,
      provider,
    });

    context('user did not approve the relayer', () => {
      let authorisation = '';

      before(async () => {
        await reset('https://rpc.ankr.com/eth', provider, 16950000);
        signer = provider.getSigner();
        address = await signer.getAddress();

        await setTokenBalance(
          signer,
          '0xa6468eca7633246dcb24e5599681767d27d1f978',
          1,
          '1000000000000000000',
          true
        );
        authorisation = await Relayer.signRelayerApproval(
          relayerAddress,
          address,
          signer,
          vault
        );
      });

      it('should build gauge2gauge with signed authorisation', async () => {
        const from = '0xa6468eca7633246dcb24e5599681767d27d1f978';
        const to = '0x57ab3b673878c3feab7f8ff434c40ab004408c4c';
        const balance = await getBalance(from, to);

        const { runGauge2Gauge } = setupMigrations(migrations, signer);

        const balanceAfter = await runGauge2Gauge(
          address,
          from,
          to,
          balance,
          authorisation
        );

        expect(balanceAfter).to.be.eql(balance);
      });
    });

    context('user approved the relayer', () => {
      beforeEach(async () => {
        await reset('https://rpc.ankr.com/eth', provider, 16950000);
        signer = await impersonateAccount(address, provider);

        // approve relayer
        await vault
          .connect(signer)
          .setRelayerApproval(address, relayerAddress, true);
      });

      context('Metastable to Metastable', () => {
        const from = metaStable;
        const to = from;

        describe('bptHodler', () => {
          before(() => {
            address = '0x21ac89788d52070D23B8EaCEcBD3Dc544178DC60';
          });

          it('joins a new pool with an limit', async () => {
            const { peekPool2Pool, runPool2Pool } = setupMigrations(
              migrations,
              signer
            );

            const { balance, expectedBptOut } = await peekPool2Pool(
              address,
              from,
              to
            );

            const balanceAfter = await runPool2Pool(
              address,
              from,
              to,
              balance,
              expectedBptOut
            );

            expect(balanceAfter).to.be.eq(expectedBptOut);
          });
        });

        describe('staked bpt', () => {
          before(() => {
            address = '0xe8343fd029561289CF7359175EE84DA121817C71';
          });

          it('should build a migration using exit / join and stake tokens in the gauge', async () => {
            const gauge = (await gaugesRepository.findBy(
              'poolId',
              from.id
            )) as {
              id: string;
            };
            const balance = await getBalance(gauge.id, address);

            const peek = await migrations.pool2poolWithGauges({
              user: address,
              from: from.id,
              to: to.id,
              balance,
            });
            const expectedBptOut = await peekBptOut(signer, peek);

            const txParams = await migrations.pool2poolWithGauges({
              user: address,
              from: from.id,
              to: to.id,
              balance,
              minBptOut: expectedBptOut,
            });

            await sendTransaction(signer, txParams);

            const balanceAfter = await getBalance(gauge.id, address);

            expect(balanceAfter).to.be.eq(expectedBptOut);
          });
        });
      });

      context('ComposableStable to ComposableStable', () => {
        before(() => {
          address = '0x74C3646ADad7e196102D1fE35267aDFD401A568b';
        });

        it('should build a migration using exit / join', async () => {
          const pool = composableStable;

          const { peekPool2Pool, runPool2Pool } = setupMigrations(
            migrations,
            signer
          );
          const { balance, expectedBptOut } = await peekPool2Pool(
            address,
            pool,
            pool
          );

          // NOTICE: When swapping from Linear Pools, the swap will query for the current wrapped token rate.
          // It is possible that the rate changes between the static call checking for the BPT out
          // and the actual swap, causing it to fail with BAL#208.
          // To avoid this, we can add a small buffer to the min BPT out amount. eg. 0.0000001% of the BPT amount.
          const buffer = BigInt(expectedBptOut) / BigInt(1e9);
          const minBptOut = String(BigInt(expectedBptOut) - buffer);

          const balanceAfter = await runPool2Pool(
            address,
            pool,
            pool,
            balance,
            minBptOut
          );

          // NOTICE: We don't know the exact amount of BPT that will be minted,
          // because swaps from the linear pool are not deterministic due to external rates
          expect(BigInt(balanceAfter)).to.satisfy(
            (v: bigint) => v > v - v / buffer && v < v + v / buffer
          );
        });
      });

      context('Weighted to Weighted between different pools', () => {
        before(() => {
          address = '0x673CA7d2faEB3c02c4cDB9383344ae5c9738945e';
        });

        it('should build a migration using exit / join', async () => {
          const from = vitaDao1;
          const to = vitaDao2;
          const { peekPool2Pool, runPool2Pool } = setupMigrations(
            migrations,
            signer
          );
          const { balance, expectedBptOut } = await peekPool2Pool(
            address,
            from,
            to
          );

          const balanceAfter = await runPool2Pool(
            address,
            from,
            to,
            balance,
            expectedBptOut
          );

          expect(balanceAfter).to.be.eq(expectedBptOut);
        });
      });

      context('gauge to gauge', () => {
        before(() => {
          address = '0xaF297deC752c909092A117A932A8cA4AaaFF9795';
        });

        it('should build a migration using exit / join and stake tokens in the gauge', async () => {
          const from = '0xa6468eca7633246dcb24e5599681767d27d1f978';
          const to = '0x57ab3b673878c3feab7f8ff434c40ab004408c4c';
          const balance = await getBalance(from, to);

          const { runGauge2Gauge } = setupMigrations(migrations, signer);

          const balanceAfter = await runGauge2Gauge(address, from, to, balance);

          expect(balanceAfter).to.be.eql(balance);
        });
      });
    });

    context('polygon', () => {
      const {
        addresses: { contracts },
      } = BALANCER_NETWORK_CONFIG[137];
      const relayerAddress = contracts.relayer;
      const provider = new JsonRpcProvider('http://127.0.0.1:8137');
      const vault = Vault__factory.connect(contracts.vault, provider);
      let signer: JsonRpcSigner;
      let address: string;

      const migrations = new Migrations({
        relayerAddress,
        poolsRepository: polygonPoolRepository,
        gaugesRepository,
        provider,
      });

      beforeEach(async () => {
        await reset('https://rpc.ankr.com/polygon', provider, 42462957);
        signer = await impersonateAccount(address, provider);

        // approve relayer
        await vault
          .connect(signer)
          .setRelayerApproval(address, relayerAddress, true);
      });

      context('ComposableStable to ComposableStable', () => {
        before(() => {
          address = '0xe80a6a7b4fdadf0aa59f3f669a8d394d1d4da86b';
        });

        it.skip('should build a migration using exit / join', async () => {
          const pool = polygonComposableStable;

          const { peekPool2Pool, runPool2Pool } = setupMigrations(
            migrations,
            signer
          );
          const { balance, expectedBptOut } = await peekPool2Pool(
            address,
            pool,
            pool
          );

          // NOTICE: When swapping from Linear Pools, the swap will query for the current wrapped token rate.
          // It is possible that the rate changes between the static call checking for the BPT out
          // and the actual swap, causing it to fail with BAL#208.
          // To avoid this, we can add a small buffer to the min BPT out amount. eg. 0.0000001% of the BPT amount.
          const buffer = BigInt(expectedBptOut) / BigInt(1e14); // 0.0000001%
          const minBptOut = String(BigInt(expectedBptOut) - buffer);

          const balanceAfter = await runPool2Pool(
            address,
            pool,
            pool,
            balance,
            minBptOut
          );

          // NOTICE: We don't know the exact amount of BPT that will be minted,
          // because swaps from the linear pool are not deterministic due to external rates
          expect(BigInt(balanceAfter)).to.satisfy(
            (v: bigint) => v > v - buffer && v < v + buffer
          );
        });
      });
    });
  });
});
```
Note that I used a `setupMigrations` helper to avoid passing `migrations` and `signer` parameters all the time.

@brunoguerios I think that testFlow is a great idea to quickly write and read inputs vs expectations but the problem is that it makes it more difficult for the developer to:
1 . Read and understand the intention of each concrete test in isolation
2. Skip or focus a single test with testing framework tools such as `it.only` or `it.skip` (as the describe/it blocks are hidden inside testFlow function).
3. Adding an edge case condition/expectation to an specific test is more difficult because everything is abstracted into the same generic function (testFlow) instead of having explicit composable helper functions in each test.

Reply:
Btw, thanks for the detailed feedback @agualis! It's very helpful 😊

Reply:
I like the purpose specific functions like getBalances etc.

Reply:
I updated the spec taking inspiration with mixing purpose specific functions and completely extracting business logic to make it super easy to follow what is being tested. I agree with @brunoguerios that the flows are always the same, so makes sense to not repeat them in the spec body.